carla/recourse_methods/catalog/roar/model.py

```diff
@@ -0,0 +1,273 @@
+        self._sample_around_instance = checked_hyperparams["sample"]
+        self._lime_seed = checked_hyperparams["lime_seed"]
+
+    def _get_lime_coefficients(
+        self, factuals: pd.DataFrame
+    ) -> Tuple[np.ndarray, np.ndarray]:
+        """
+        ROAR Recourse is only defined on linear models. To make it work for arbitrary non-linear networks
+        we need to find the lime coefficients for every instance.
+
+        Parameters
+        ----------
+        factuals : pd.DataFrame
+            Instances we want to get lime coefficients
+
+        Returns
+        -------
+        coeffs : np.ndArray
+        intercepts : np.ndArray
+
+        """
+
+        np.random.seed(self._lime_seed)
+
+        coeffs = np.zeros(factuals.shape)
+        intercepts = []
+        lime_data = self._data.df[self._mlmodel.feature_input_order]
+        lime_label = self._data.df[self._data.target]
+
+        lime_exp = LimeTabularExplainer(
+            training_data=lime_data.values,
+            training_labels=lime_label,
+            feature_names=self._mlmodel.feature_input_order,
+            discretize_continuous=self._discretize_continuous,
+            sample_around_instance=self._sample_around_instance,
+            categorical_names=[
+                cat
+                for cat in self._mlmodel.feature_input_order
+                if cat not in self._data.continuous
+            ]
+            # self._data.encoded_normalized's categorical features contain feature name and value, separated by '_'
+            # while self._data.categorical do not contain those additional values.
+        )
+
+        for index, row in factuals.iterrows():
+            factual = row.values
+            explanations = lime_exp.explain_instance(
+                factual,
+                self._mlmodel.predict_proba,
+                num_features=len(self._mlmodel.feature_input_order),
+            )
+            intercepts.append(explanations.intercept[1])
+
+            for tpl in explanations.local_exp[1]:
+                coeffs[index][tpl[0]] = tpl[1]
+
+        return coeffs, np.array(intercepts)
+
+    def get_counterfactuals(self, factuals: pd.DataFrame) -> pd.DataFrame:
+        factuals = factuals.reset_index()
+        factuals = self._mlmodel.get_ordered_features(factuals)
+
+        encoded_feature_names = self._mlmodel.data.encoder.get_feature_names(
+            self._mlmodel.data.categorical
+        )
+        cat_features_indices = [
+            factuals.columns.get_loc(feature) for feature in encoded_feature_names
+        ]
+
+        coeffs = self._coeffs
+        intercepts = self._intercepts
+
+        # Calculate coefficients and intercept (if not given) and reshape to match the shape that LIME generates
+        # If Model linear then extract coefficients and intercepts from raw model directly
+        # If Molel ANN then use LIME to generate the coefficients
+        if (coeffs is not None) and (intercepts is not None):
```

Comment:
I think it's better to switch the if-else, such that you get: if `coeffs is None and/or intercepts is None` - else.